specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1079 @@
+    ResponseMeta
+    // PullLocked requests for locked state needed to perform atomic exec.
+    PullLocked
+    // ResponseLocked is used to answer to pull requests for locked state.
+    ResponseLocked
+    // PullCheck requests the checkpoint for subnet. This request
+    // needs to be sent to the parent of a subnet (which is the one)
+    // persisting the committed checkpoints for subnets.
+    PullCheck 
+    // ResponseCheck answers a request for a checkpoint pull request.
+    ResponseCheck
+)
+```
+
+> Resolve message data structure
+```go
+type ResolveMsg struct {
+	// From subnet
+	From address.SubnetID
+	// Message type being propagated
+	Type MsgType
+	// Cid of the content
+	Cid cid.Cid
+	// MsgMeta being propagated (if any)
+	CrossMsgs CrossMsgs
+	// Checkpoint being propagated (if any)
+	Checkpoint Checkpoint
+	// LockedState being propagated (if any).
+	Locked atomic.LockedState
+    // address wrapped as string to support undef serialization
+	Actor  string 
+}
+```
+
+![](https://hackmd.io/_uploads/rJd0_6tK9.png)
+
+### Data availability
+Peers requesting content from others subnets trust that the peers in the other network from which a checkpoint with cross-msgs has been propagated will resolve the content resolution requests sent to the subnet. Intuitively, the node that triggered the cross-msg has no incentive on denying access to this data (as his funds have already been burnt), but data availability is an issue that is worth addressing further. The aforementioned design of the content resolution protocol assumes at least _one honest participant in the subnet_ (i.e. a peer that always answers successfully to content resolution requests) _and that the data in the subnet is always available_. In a real environment these assumption may not hold, and more complex schemes may be needed to incentivize subnet peers and ensure that every content resolution request between subnets is fulfilled, and a high-level of availability of data.
+
+To overcome this issue, `SCA` in subnets include the `Save()` function and peers will implement a protocol to backup the state of the function in any storage system where data can be retrievable and available independently of the state of a subnet (let this be Filecoin storage, IPFS, or any other decentralized or centralized storage systems). Having the state always available is key for:
+- The execution and validations of cross-net messages.
+- Creating fraud/fault proofs from detectable misbehaviors in a subnet.
+- Migrating the state of a subnet and spawning a new subnet from the existing state of another network.
+
+> TODO: For this puropse, a `Persistence` interface will be explored and implemented in future iterations of the protocol. Ideally we should piggy-back from all the available storage in the Filecoin network, FVM native integration with storage primitives, and all the data retrievable work being done by CryptoNetLab.
```

Comment:
data retrievabillity (typo), give link to CryptoNetLab work (as it is unclear from here) and also perhaps mention Filecoin Saturn work (retrieval.market)